### PR TITLE
remove pure variables

### PR DESCRIPTION
**Pull request by AndHager**
Implemented a function that removes at the begin of cdcl all pure variables and assigns values to them.
This modification improves the performance of the cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_135_510.cnf bench from around 700  to 5 seconds.

### CODE REVIEW

File: src/CDCL/Types.hs [hunk 1]
```diff
@@ -74,7 +74,11 @@ data BoolVal =
     |
         -- | Valuewise -1
         BNothing
-    deriving (Show, Eq, Ord)
+    deriving (Eq, Ord)
+instance Show BoolVal where
+    show BFalse = "False"
+    show BTrue = "True"
+    show BNothing = "Noting"
```
Comment (mgudemann) [inline src/CDCL/Types.hs]: please revert, unrelated change

File: src/CDCL/Types.hs [hunk 2]
```diff
@@ -90,7 +94,10 @@ data InterpretResult =
 
 -- | Literal defined as Integer
 newtype Literal = Lit Integer
-    deriving (Show, Eq, Ord)
+    deriving (Eq, Ord)
+instance Show Literal where
+    show (Lit i) = "'" ++ show i ++ "'"
+
```
Comment (mgudemann) [inline src/CDCL/Types.hs]: please revert, unrelated change

File: src/CDCL/Algorithm.hs [hunk 3]
```diff
@@ -35,18 +35,22 @@ import           CDCL.Unitpropagation (unitPropagation, unitResolution,
...
-
+import           CDCL.DPLL (rmPureVars)
+ 
 import qualified Data.Map.Strict as Map
 import           Data.Maybe
 
 hardCoded = Period 30
 startBoundary = 20
 
-
```
Comment (mgudemann) [inline src/CDCL/Algorithm.hs]: please remove whitespace changes

File: src/CDCL/Algorithm.hs [hunk 4]
```diff
@@ -58,14 +62,26 @@ solve t = case result of
...
 --   Function will immediately return SAT if ClauseList is null or UNSAT if an empty List is found within clist
 cdcl :: [[Integer]] -> Bool -> Bool -> CDCLResult
-cdcl clist stats fullStats
+cdcl clist stats fullStats = case result of 
+            SAT l                             -> SAT (l ++ decodedPureVals)
+            SAT_WITH_STATS l a0 a1 a2 a3         -> SAT_WITH_STATS (l ++ decodedPureVals) a0 a1 a2 a3
+            SAT_WITH_FULL_STATS l a0 a1 a2 a3 a4 a5 -> SAT_WITH_FULL_STATS (l ++ decodedPureVals) a0 a1 a2 a3 a4 a5
+            UNSAT                             -> UNSAT
+            UNSAT_WITH_STATS a0 a1            -> UNSAT_WITH_STATS a0 a1
+            where (reducedTerm, pureVars) = rmPureVars clist
```
Comment (mgudemann) [inline src/CDCL/Algorithm.hs]: please align all `->`

File: src/CDCL/Algorithm.hs [hunk 5]
```diff
@@ -101,28 +117,38 @@ cdcl clist stats fullStats
...
+  -> ClauseList -- clistOG : Clause List in Original Form
+  -> ClauseList -- learnedClist
+  -> [Clause] -- learnedClauses
+  -> [Clause] -- confClauses
+  -> ClauseList -- clist : ClauseList :: [
+                --                          ReducedClauseAndOGClause :: (
+                --                              Clause :: [Literal :: Lit Integer]    -- Clause reduced with Unitresolution
+                --                              , Clause :: [Literal :: Lit Integer]  -- Clause in its original form
+                --                          )
+                --                       ]
```
Comment (mgudemann) [inline src/CDCL/Algorithm.hs]: please keep to line comments in type signature

File: src/CDCL/Algorithm.hs [hunk 6]
```diff
@@ -101,28 +117,38 @@ cdcl clist stats fullStats
...
+             --                         if conflictIteration == upperBoundary
+             --                             or conflictIteration == currentBoundary
+             --                         then the algorithm will be restarted 
+             --                             with higher upperBoundary currentBoundary. 
+  -> Integer -- upperBoundary
+  -> Integer -- currentBoundary
+  -> Bool -- stats
+  -> Bool -- fullStats
+  -> Integer -- restarts
+  -> CDCLResult -- result
```
Comment (mgudemann) [inline src/CDCL/Algorithm.hs]: please use haddock style documentation for type signature, i.e.,
```
-> Int -- ^ The 'Int' argument
-> ...
```

File: src/CDCL/DPLL.hs [hunk 7]
```diff
@@ -0,0 +1,50 @@
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
+
+module CDCL.DPLL (rmPureVars, getPureVars, isPureVar, rmVars) where
+
+import Data.Set
+
+rmPureVars :: [[Integer]] -> ([[Integer]], [Int])
```
Comment (mgudemann) [inline src/CDCL/DPLL.hs]: please add haddock documentation for functions (at least for exported ones)

File: src/CDCL/DPLL.hs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+getPureVars'' fullerm (x:xs) = [i | isPureVar fullerm x] ++ getPureVars'' fullerm xs
+    where i = fromIntegral x
+
+isPureVar :: [[Integer]] -> Integer -> Bool
+isPureVar [] _ = True
+isPureVar (x:xs) var = (not (containsVar x (-var))) && isPureVar xs var
+
+containsVar :: [Integer] -> Integer -> Bool
+containsVar [] _ = False
+containsVar (x:xs) var = (x == var) || containsVar xs var
```
Comment (mgudemann) [inline src/CDCL/DPLL.hs]: use `Data.List.notElem` instead

**Review by mgudemann — CHANGES_REQUESTED** [review]

**Comment by mgudemann** [comment]
please rebase onto `main`

**Comment by mgudemann** [comment]
#2 removed output of satisfying valuation, reducing the size of output for larger SAT instances
Could you provide some more benchmarks for `opt` after rebase? `stack test` is slower with `opt` but that may be irrelevant.

**Comment by AndHager** [comment]
Benchmarks of the unmodified sat-solver:
Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_10_17.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_102_481.cnf
Result: SAT
CPU time:   1.67s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_108_345.cnf
Result: SAT
CPU time:   0.28s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_112_238.cnf
Result: SAT
CPU time:   0.07s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_117_328.cnf
Result: SAT
CPU time:   0.35s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_120_1016.cnf
Result: SAT
CPU time:  16.30s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_120_724.cnf
Result: SAT
CPU time:   0.86s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_12_16.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_130_393.cnf
Result: SAT
CPU time:   0.32s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_135_510.cnf
Result: SAT
CPU time: 700.95s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_140_1371.cnf
Result: SAT
CPU time:   1.99s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_14_19.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_144_648.cnf
Result: SAT
CPU time:   2.84s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_15_23.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_153_714.cnf
Result: SAT
CPU time:   0.22s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_160_1028.cnf
Result: SAT
CPU time:  10.80s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_16_24.cnf
Result: SAT
CPU time:   0.01s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_168_722.cnf
Result: SAT
CPU time:   3.01s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_170_1247.cnf
Result: SAT
CPU time:   3.34s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_18_21.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_18_49.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_185_1132.cnf
Result: SAT
CPU time:  13.91s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_195_899.cnf
Result: SAT
CPU time:   8.78s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_196_1092.cnf
Result: SAT
CPU time: 295.45s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_210_1275.cnf
Result: SAT
CPU time:   8.52s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_21_40.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_222_1477.cnf
Result: SAT
CPU time:  60.85s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_228_3437.cnf
Result: SAT
CPU time:  10.39s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_24_34.cnf
Result: SAT
CPU time:   0.05s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_24_56.cnf
Result: SAT
CPU time:   0.04s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_252_3132.cnf
Result: SAT
CPU time: 6166.78s

**Comment by AndHager** [comment]
Benchmarks of the modified sat-solver:
Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_10_17.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_102_481.cnf
Result: SAT
CPU time:   0.22s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_108_345.cnf
Result: SAT
CPU time:   0.09s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_112_238.cnf
Result: SAT
CPU time:   0.17s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_117_328.cnf
Result: SAT
CPU time:   0.12s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_120_1016.cnf
Result: SAT
CPU time:  54.91s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_120_724.cnf
Result: SAT
CPU time:   2.26s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_12_16.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_130_393.cnf
Result: SAT
CPU time:   0.09s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_135_510.cnf
Result: SAT
CPU time:   4.61s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_140_1371.cnf
Result: SAT
CPU time:   1.12s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_14_19.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_144_648.cnf
Result: SAT
CPU time:   0.34s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_15_23.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_153_714.cnf
Result: SAT
CPU time:   0.60s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_160_1028.cnf
Result: SAT
CPU time:   0.46s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_16_24.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_168_722.cnf
Result: SAT
CPU time:   0.39s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_170_1247.cnf
Result: SAT
CPU time:   0.97s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_18_21.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_18_49.cnf
Result: SAT
CPU time:   0.16s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_185_1132.cnf
Result: SAT
CPU time:   9.74s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_195_899.cnf
Result: SAT
CPU time:   0.60s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_196_1092.cnf
Result: SAT
CPU time:  46.84s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_210_1275.cnf
Result: SAT
CPU time:   0.23s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_21_40.cnf
Result: SAT
CPU time:   0.16s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_222_1477.cnf
Result: SAT
CPU time:  53.41s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_228_3437.cnf
Result: SAT
CPU time:   3.19s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_24_34.cnf
Result: SAT
CPU time:   0.00s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_24_56.cnf
Result: SAT
CPU time:   0.03s

Executing: cadp.inria.fr/ftp/benchmarks/vlsat/vlsat1_252_3132.cnf
Result: SAT
CPU time:  72.82s

**Comment by mgudemann** [comment]
@AndHager LGTM

#3 seems to be a fallout from this

**Pull request merged**